Improve building ASP.NET Core directly on arm64 machines

omajid opened the pull request:
[This is still a work in progress]

This PR tries to get ASP.NET Core repo to build more easily and tests to pass on Arm64. I am only testing on Linux, though. Please note that unlike the CI environments which build and test Arm64 while running on (and cross-compiling on) x64, my goal here is to get this to work on Arm64 machines themselves.

My goal is to get `eng/build.sh --arch arm64 --test` to fully pass on arm64. Failing that, I would like to get to the same set of test failures that I see on x64.

omajid committed:
Improve build on Linux arm64

TanayParikh commented on Directory.Build.props:
```diff
@@ -172,6 +172,10 @@
       freebsd-x64
     </SupportedRuntimeIdentifiers>
 
+    <!-- Playwright provides binaries for Windows (x86 and x64), macOS (x64) and Linux (x64, non musl). We can't use it on other architectures. -->
+    <IsPlaywrightAvailable Condition="'$(TargetOsName)' == 'linux-musl' OR ('$(TargetArchitecture)' != 'x86' AND '$(TargetArchitecture)' != 'x64')">false</IsPlaywrightAvailable>
+    <IsPlaywrightAvailable Condition="'$(IsPlaywrightAvailable)' == ''">true</IsPlaywrightAvailable>
```
Just cross-linking given the Playwright changes: https://github.com/dotnet/aspnetcore/pull/38414

omajid replied:
Thanks! Looking at https://nuget.info/packages/Microsoft.Playwright/1.17.0, it looks like it has even fewer os/arch combinations: Linux, macOS, and Windows all x64 only.

dougbu replied:
/fyi if I'm correct and we only use Playwright in our BlazorTemplates.Tests, then we currently use Playwright only on macOS and Windows in the CI. Testing on Linux x64 (or at least Ubuntu 18.04 / 20.04 x64) would be nice to have but I don't know if that even works locally (@TanayParikh ❔)

TanayParikh replied:
> Testing on Linux x64 (or at least Ubuntu 18.04 / 20.04 x64) would be nice to have but I don't know if that even works locally (@TanayParikh ❔)

I'm currently upgrading our Playwright usage from `v0.192.0` to `v1.16.1` via https://github.com/dotnet/aspnetcore/pull/38414. This involves a bunch of breaking changes, so currently the tests don't really work anywhere. Once that's all up and running again, I'm anticipating there shouldn't be too big of an issue running the tests on linux (along with Windows and macOS).

Side note; it looks like `v1.17.0` just got released so will look into updating my PR to use that version.

dougbu commented on global.json:
```diff
@@ -5,12 +5,11 @@
   "tools": {
     "dotnet": "7.0.100-alpha.1.21558.2",
     "runtimes": {
```
@Tratcher could you please have a look at the Server Comparison test failures❔ They probably relate to the missing .dotnet/x86 folder on Windows build agents w/ the changes in this file. But, I'd appreciate your thoughts.

Tratcher replied:
Yes, it makes sense these tests would fail without the x86 runtime. I'm a little surprised most of them gracefully fell back to x64 and only failed checking the output. Those are the easiest variations to remove. The one that didn't gracefully fall back is the one we should preserve somehow, since x86 actually mattered in that scenario.
`HelloWorld(variant: Server: IISExpress, TFM: net7.0, Type: Portable, Arch: x86, Host: InProcess)`

dougbu replied:
Let's table this unless we find we _must_ remove the "dotnet/x86" entry at some point in the future.

dougbu commented on global.json:
```diff
@@ -5,12 +5,11 @@
-      "dotnet/x86": [
-        "$(MicrosoftNETCoreBrowserDebugHostTransportVersion)"
+      "dotnet/x64": [
+        "2.1.30"
```
I'm not positive but suspect we have scripts etc. (like activate.*) that expect .dotnet/x86 for some of our special-case testing. Nothing should be looking for .dotnet/x64 and I think that's where 2.1.30 will be installed w/ this setup.

Might be worth trying

``` json
"runtimes": {
  "dotnet": [
    "2.1.30",
    "$(MicrosoftNETCoreBrowserDebugHostTransportVersion)"
   ],
  "dotnet/x86": [
    "$(MicrosoftNETCoreBrowserDebugHostTransportVersion)"
  ],
  "aspnetcore": [
    "3.1.21"
  ]
}
```

This suggestion should install all shared framework bits for the current system (Arm[64] / x64 or whatever) and add the latest Microsoft.NETCore.App for x86. But, it may not work on Arm64 machines because x86 emulation isn't available (right ❔). We don't have a way to say "please add x86 bits if this is an x64 machine" ☹️

But, maybe we can stop testing x86 on build agents. @Tratcher❔

omajid replied:
I think I completely misunderstood what this part of global.json does. Are there some docs for what this `runtimes` section is used for?

dougbu replied:
https://github.com/dotnet/arcade/blob/56bac3e40e405eacba76d048aaff104c01f4d8ee/Documentation/ArcadeSdk.md#globaljson covers some of it.

But, I checked deeper and the infrastructure we use from dotnet/arcade
1. Ignores `dotnet/x86` entries unless on Windows
2. Does **not** ignore `dotnet/x86` entries on Windows Arm64 (☹️ but building there is less of a priority)
3. Doesn't seem to change the installation location for `dotnet/x64` entries on Windows
4. Adds the specified architecture (the bit after the slash) to the installation location on Linux. But, I very much doubt we attempt to do x64 testing on Arm64 build agents or in Arm64 local builds.
5. Defaults to using the current `MSBuild` `$(Platform)` (ignoring `AnyCpu`) when no architecture portion (the bit after the slash) is specified. This may be a problem for building on Arm64.

Therefore, my suggestion might be fine because it maintains `dotnet/x86` and avoids per-x64 downloads on Arm64 build agents. Might not need the `dotnet/x86` bit if @Tratcher thinks we can eliminate x86 testing on Windows.

But, please confirm a local build on Arm64 downloads the Arm64 version of Microsoft.NETCore.App. We may need to get some help here from the Arcade team.

Tratcher replied:
x64 is far more common than x86, and I'm not are of any x86 specific issues we've had. I would hope there's still at least one x86 test run somewhere, even if its on CTI.

dougbu replied:
@omajid please try my suggestion because we need to know if leaving the `dotnet/x86` entry here breaks the Linux Arm64 build. If it doesn't, that change is easy and preserves the Server Comparison tests on Windows.

omajid replied:
@dougbu Your suggestions for global.json works for me!

dougbu replied:
Please update your branch and push to your fork. Then, we can confirm the suggestion doesn't break other bits 😃

Separately, are your local builds looking solid enough that your next iteration should be ready for review❔

dougbu commented:
> my goal here is to get this to work on Arm64 machines themselves.

👍

dougbu commented:
@omajid please comment when you think this PR is ready for review. I don't think the `Ready for review` button will notify PR subscribers.

Separately, is local testing on Arm64 Linux machines now working fine❔

omajid committed:
Apply changes suggested by Doug Bunting

omajid commented:
On an arm64 machine, this is the set of errors (and test failures) now:

```
/home/omajid/aspnetcore/eng/targets/Npm.Common.targets(79,5): error : Waiting for the other yarn instance to finish (610076) [/home/omajid/aspnetcore/src/JSInterop/Microsoft.JSInterop.JS/src/Microsoft.JSInterop.JS.npmproj]
/home/omajid/aspnetcore/eng/targets/Npm.Common.targets(79,5): error : Waiting for the other yarn instance to finish (610105) [/home/omajid/aspnetcore/src/SignalR/clients/ts/signalr/signalr.npmproj]
XUnit : error : Tests failed: /home/omajid/aspnetcore/artifacts/TestResults/Debug/Microsoft.AspNetCore.Server.Kestrel.Core.Tests_net7.0_x64.html [net7.0|x64] [/home/omajid/aspnetcore/src/Servers/Kestrel/Core/test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests.csproj]
XUnit : error : Tests failed: /home/omajid/aspnetcore/artifacts/TestResults/Debug/Interop.FunctionalTests_net7.0_x64.html [net7.0|x64] [/home/omajid/aspnetcore/src/Servers/Kestrel/test/Interop.FunctionalTests/Interop.FunctionalTests.csproj]
XUnit : error : Tests failed: /home/omajid/aspnetcore/artifacts/TestResults/Debug/Microsoft.AspNetCore.Hosting.FunctionalTests_net7.0_x64.html [net7.0|x64] [/home/omajid/aspnetcore/src/Hosting/test/FunctionalTests/Microsoft.AspNetCore.Hosting.FunctionalTests.csproj]
XUnit : error : Tests failed: /home/omajid/aspnetcore/artifacts/TestResults/Debug/ServerComparison.FunctionalTests_net7.0_x64.html [net7.0|x64] [/home/omajid/aspnetcore/src/Servers/test/FunctionalTests/ServerComparison.FunctionalTests.csproj]
```

omajid commented:
Does anyone know why `TestMatrix` likes to treat arm64 as x64? 

https://github.com/dotnet/aspnetcore/pull/27838/files#diff-9ca0861246bf9e147ad6306475ac44117462c4539754595f4c27db51bdb10d40

omajid committed:
Improve running tests on arm64

Try and use the currently running system's architecture instead of hard-coding
x64.

omajid commented:
After fixing up my build command to be `eng/build.sh --arch arm64 --restore --build --pack --test`, I am now seeing 2 remaining test failures:


```xml
      <test name="Microsoft.AspNetCore.Server.Kestrel.Core.Tests.HeartbeatTests.HeartbeatTakingLongerThanIntervalIsLoggedAsWarning" type="Microsoft.AspNetCore.Server.Kestrel.Core.Tests.HeartbeatTests" method="HeartbeatTakingLongerThanIntervalIsLoggedAsWarning" time="7.0456688" result="Fail">
        <output><![CDATA[| [0.001s] TestLifetime Information: Starting test HeartbeatTakingLongerThanIntervalIsLoggedAsWarning at 2021-11-30T14:26:05
| [7.104s] Microsoft.AspNetCore.Server.Kestrel.Core.Tests.HeartbeatTests Error: Test threw an exception.
| System.TimeoutException: The operation has timed out.
|    at Microsoft.AspNetCore.Testing.TaskExtensions.TimeoutAfter(Task task, TimeSpan timeout, String filePath, Int32 lineNumber) in /home/omajid/aspnetcore/src/Shared/TaskExtensions.cs:line 125
|    at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.HeartbeatTests.HeartbeatTakingLongerThanIntervalIsLoggedAsWarning() in /home/omajid/aspnetcore/src/Servers/Kestrel/Core/test/HeartbeatTests.cs:line 50
|    at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
| --- End of stack trace from previous location ---
|    at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
|    at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
| [7.111s] TestLifetime Information: Finished test HeartbeatTakingLongerThanIntervalIsLoggedAsWarning in 7.1092629s
]]></output>
        <failure exception-type="System.TimeoutException">
          <message><![CDATA[System.TimeoutException : The operation has timed out.]]></message>
          <stack-trace><![CDATA[   at Microsoft.AspNetCore.Testing.TaskExtensions.TimeoutAfter(Task task, TimeSpan timeout, String filePath, Int32 lineNumber) in /home/omajid/aspnetcore/src/Shared/TaskExtensions.cs:line 125
   at Microsoft.AspNetCore.Server.Kestrel.Core.Tests.HeartbeatTests.HeartbeatTakingLongerThanIntervalIsLoggedAsWarning() in /home/omajid/aspnetcore/src/Servers/Kestrel/Core/test/HeartbeatTests.cs:line 50
--- End of stack trace from previous location ---]]></stack-trace>
        </failure>
      </test>
```

I am seeing this on x64 as well, so maybe it's my configuration?


```xml
      <test name="Interop.FunctionalTests.H2SpecTests.RunIndividualTestCase" type="Interop.FunctionalTests.H2SpecTests" method="RunIndividualTestCase" time="0" result="Fail">
        <failure exception-type="System.NotImplementedException">
          <message><![CDATA[System.NotImplementedException : Invalid OS]]></message>
          <stack-trace><![CDATA[   at Interop.FunctionalTests.H2SpecCommands.GetToolLocation() in /home/omajid/aspnetcore/src/Servers/Kestrel/test/Interop.FunctionalTests/H2SpecCommands.cs:line 65
   at Interop.FunctionalTests.H2SpecCommands.EnumerateTestCases() in /home/omajid/aspnetcore/src/Servers/Kestrel/test/Interop.FunctionalTests/H2SpecCommands.cs:line 71
   at Interop.FunctionalTests.H2SpecTests.get_H2SpecTestCases() in /home/omajid/aspnetcore/src/Servers/Kestrel/test/Interop.FunctionalTests/H2SpecTests.cs:line 65]]></stack-trace>
        </failure>
      </test>
```

I see that the project file says to skip this on Helix Arm64. Is there some way to say "skip this on arm64", and rely on that to skip for helix too?

dougbu commented:
> Is there some way to say "skip this on arm64", and rely on that to skip for helix too?

If the check should be in C#, add an attribute like [`[SkipOnAlpine]`](https://github.com/dotnet/aspnetcore/blob/main/src/Testing/src/xunit/SkipOnAlpineAttribute.cs) or [`[SkipOn32BitOS]`](https://github.com/dotnet/aspnetcore/blob/main/src/Hosting/Server.IntegrationTesting/src/xunit/SkipOn32BitOSAttribute.cs) in either src/Testing/src/xunit/ or the project containing the relevant tests (if just one).

If the check should be in MSBuild, `<SkipTests Condition=" '$(TargetArchitecture)' == 'arm64' ">true</SkipTests>` should do the trick when testing locally. This does _not_ cover Helix runs, where `<SkipHelixArm>true</SkipHelixArm>` handles our testing fan-out.

omajid committed:
Skip h2spec tests on arm64 and x86

omajid commented:
Some tests are still flaky on my end, but now I am down to:

```
/home/omajid/aspnetcore/eng/targets/Npm.Common.targets(79,5): error : Waiting for the other yarn instance to finish (2
545773) [/home/omajid/aspnetcore/src/JSInterop/Microsoft.JSInterop.JS/src/Microsoft.JSInterop.JS.npmproj]
/home/omajid/aspnetcore/eng/targets/Npm.Common.targets(79,5): error : Waiting for the other yarn instance to finish (2
545789) [/home/omajid/aspnetcore/src/SignalR/clients/ts/signalr/signalr.npmproj]                                      
    0 Warning(s)                                           
    2 Error(s) 
```

These aren't even real errors, right?

dougbu commented:
> These aren't even real errors, right?

Right. Those can be ignored and usually go away on retry.

dougbu commented on src/Hosting/Server.IntegrationTesting/src/TestMatrix.cs:
```diff
@@ -53,10 +53,17 @@ public TestMatrix WithArchitectures(params RuntimeArchitecture[] archs)
         return this;
     }
 
+    /// <summary>
+    /// With all architectures that are compatible with the currently running architecture
+    /// </summary>
+    /// <returns></returns>
```
nit: No need for doc comments in a test or test infrastructure project

omajid replied:
Sorry, PR got merged before I could remove this.

I only added this one because the method now does something *slightly* different than the name suggests. The name says "all architectures" but the method is actually "all architectures compatible with the current architecture". Changing the method name seemed overkill, so I left it as a doc comment.

dougbu commented on src/Servers/Kestrel/test/Interop.FunctionalTests/H2SpecTests.cs:
```diff
@@ -20,6 +21,8 @@ namespace Interop.FunctionalTests;
 
 public class H2SpecTests : LoggedTest
 {
+
```
nit: Suggest removing this blank line

dougbu reviewed (approved):
Looks fine. Many thanks❕

No need to do anything about the "nit" items unless another iteration is needed for some reason

dougbu commented:
thanx again❕
